src/daemon/auth/providers/openid_connect/provider.rs

```diff
@@ -357,49 +363,84 @@ impl OpenIDConnectAuthProvider {
-                                                return Some(Auth::Bearer(new_token));
+                                                // The new token was successfully acquired from the OpenID Connect Provider,
+                                                // and early returned.
+                                                return Ok(Auth::Bearer(new_token));
                                             }
                                             Err(err) => {
-                                                warn!(
-                                                    "OpenID Connect: Could not extend login session for user '{}': {}",
-                                                    &session.id, err
-                                                );
+                                                return Err(CoreErrorResponseType::Extension(format!(
+                                                    "Error while encoding the refreshed token {}",
+                                                    err
+                                                )));
                                             }
                                         }
                                     }
                                     Err(err) => {
-                                        warn!(
-                                            "OpenID Connect: Could not extend login session for user '{}': {}",
-                                            &session.id, err
-                                        );
+                                        match &err {
+                                            // this is where the RFC-6749 5.2 Error Response is received and
+                                            // return to the caller. It's the responsibility of the caller
+                                            // to decide whether to retry or report back to the user.
+                                            //
+                                            // Note that [Errata for RFC 6749](https://www.rfc-editor.org/errata/eid4745)
+                                            // defines two additional error responses, `server_error` and
+                                            // `temporarily_unavailable`, that don't have variant counterparts
+                                            // in the openid-connect crate. These two error messages will
+                                            // therefore **not** end up in the `ServerReponse` variant.
+                                            openidconnect::RequestTokenError::ServerResponse(r) => {
+                                                return Err(r.error().clone());
+                                            }
+                                            openidconnect::RequestTokenError::Request(r) => {
+                                                return Err(CoreErrorResponseType::Extension(format!(
```

Comment:
One thing I'm wondering about here is error chaining. In effect you are chaining two errors together by concatenating their error messages. However, the oauth2 crate from which RequestTokenError comes uses the failure crate and the [docs for that crate include](https://docs.rs/failure/0.1.8/failure/struct.Causes.html) this:

> Causes | A iterator over the causes of a Fail

Maybe the current approach is fine for now but I wanted to raise this in case it's really worth looking at. I have no idea.

Reply:
The ServerResponse(T) variant's T type has bounds `ErrorResponse + 'static`, so no `Fail`. It doesn't have a cause() method. The ErrorResponse which wraps the StandardErrorResponse trait, which feature the error\_description() method. According to the docs "`... is used to assist the client developer in understanding the error that occurred."`

So I don't think that should be propagated up the error chain, since it doesn't give the end user anything. We could add a `trace!` statement for it.

The other variants here are all non-RFC 6749 error messages, more generic messages, like known unknowns and unknown unknowns.

So I think the current approach is good enough for now